File: api/java/toolgood.textfilter.api/src/main/java/toolgood/textfilter/api/TextFilterConfig.java
package toolgood.textfilter.api;

import java.util.List;

import toolgood.textfilter.api.ConsulDiscovery.ConsulServiceDiscovery;
import toolgood.textfilter.api.Grpcs.TextFilterGrpcAsyncProvider;
import toolgood.textfilter.api.Grpcs.TextFilterGrpcProvider;
import toolgood.textfilter.api.Impl.KeywordProvider;
import toolgood.textfilter.api.Impl.KeywordTypeProvider;
import toolgood.textfilter.api.Impl.SysProvider;
import toolgood.textfilter.api.Impl.TextFilterAsyncProvider;
import toolgood.textfilter.api.Impl.TextFilterProvider;
import toolgood.textfilter.api.Interfaces.IKeywordProvider;
import toolgood.textfilter.api.Interfaces.IKeywordTypeProvider;
import toolgood.textfilter.api.Interfaces.ISysProvider;
import toolgood.textfilter.api.Interfaces.ITextFilterAsyncProvider;
import toolgood.textfilter.api.Interfaces.ITextFilterProvider;

public class TextFilterConfig {
    private static TextFilterConfig _textFilterConfig;

    public static TextFilterConfig Instance() {
        if (_textFilterConfig == null) {
            _textFilterConfig = new TextFilterConfig();
        }
        return _textFilterConfig;
    }

    private String _grpcHost;
    private String _textFilterHost;
    private String _registryAddress = "http://localhost:8500";

    public void SetConsulAddress(String value) {
        _registryAddress = value;
    }

    public void SetTextFilterHost(String value) {
        _textFilterHost = value;
    }

    public String GetTextFilterHost() {
        if (_textFilterHost == null) {
            return ConsulServiceDiscovery.DiscoveryOne(_registryAddress, "ToolGood.TextFilter");
        }
        return _textFilterHost;
    }

    public void SetGrpcHost(String value) {
        _grpcHost = value;
    }

    public String GetGrpcHost() {
        if (_grpcHost == null) {
            return ConsulServiceDiscovery.DiscoveryOne(_registryAddress, "ToolGood.TextFilter.Grpc");
        }
        return _grpcHost;
    }

    public IKeywordProvider CreateKeywordProvider() {
        return new KeywordProvider(this);
    }

    public IKeywordTypeProvider CreateKeywordTypeProvider() {
        return new KeywordTypeProvider(this);
    }

    public ISysProvider CreateSysProvider() {
        return new SysProvider(this);
    }

    public ITextFilterProvider CreateTextFilterProvider() {
        return new TextFilterProvider(this);
    }

    public ITextFilterAsyncProvider CreateTextFilterAsyncProvider() {
        return new TextFilterAsyncProvider(this);
    }

    public TextFilterGrpcProvider CreateTextFilterGrpcClient() {
        return new TextFilterGrpcProvider(GetGrpcHost().replace("http://", ""));
    }

    public TextFilterGrpcAsyncProvider CreateTextFilterGrpcAnsycClient() {
        return new TextFilterGrpcAsyncProvider(GetGrpcHost().replace("http://", ""));
    }

    public List<String> GetServiceUrls(ServiceUrlType urlType) {
        if (urlType == ServiceUrlType.Grpc) {
            return ConsulServiceDiscovery.Discovery(_registryAddress, "ToolGood.TextFilter.Grpc");
        }
        return ConsulServiceDiscovery.Discovery(_registryAddress, "ToolGood.TextFilter");
    }

}

File: api/csharp/ToolGood.TextFilter.Api/TextFilterConfig.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Threading.Tasks;
using ToolGood.TextFilter.Api.Impl;
using ToolGood.TextFilter.Api.Interfaces;
#if Async
using Grpc.Net.Client;
using ToolGood.TextFilter.Api.Grpcs;
#endif
#if Consul
using ToolGood.TextFilter.Api.ConsulDiscovery;
#endif



namespace ToolGood.TextFilter.Api
{
    public class TextFilterConfig
    {

        private const string HttpServiceName = "ToolGood.TextFilter";
        private const string GrpcServiceName = "ToolGood.TextFilter.Grpc";
        public static int OvertimeSeconds = 10;

        private static TextFilterConfig _textFilterConfig;
        public static TextFilterConfig Instance {
            get {
                if (_textFilterConfig == null) {
                    _textFilterConfig = new TextFilterConfig();
                }
                return _textFilterConfig;
            }
        }



        #region TextFilterHost
        private string _textFilterHost;
        private string _tempTextFilterHost;
        private DateTime _maxTemp_Text;

        /// <summary>
        /// 不设置  取Consul
        /// </summary>
        public string TextFilterHost {
            set {
                _textFilterHost = value;
            }
            get {
#if Consul
                if (_textFilterHost == null) {
                    if (string.IsNullOrEmpty(_tempTextFilterHost) || DateTime.Now > _maxTemp_Text) {
                        var task = ConsulServiceDiscovery.DiscoveryOne(ConsulAddress, HttpServiceName);
                        task.Wait();
                        _tempTextFilterHost = task.Result;
                        _maxTemp_Text = DateTime.Now.AddSeconds(OvertimeSeconds);
                    }
                    return _tempTextFilterHost;
                }
#endif
                return _textFilterHost;
            }
        }
        #endregion

        public Func<HttpClient> SetHttpClientFunc { private get; set; }
        public IHttpClientFactory HttpClientFactory { private get; set; }
        public string FactoryName { private get; set; }
        internal HttpClient CreateHttpClient()
        {
            if (HttpClientFactory != null) {
                return HttpClientFactory.CreateClient(FactoryName ?? "");
            }
            if (SetHttpClientFunc != null) {
                return SetHttpClientFunc();
            }
            return new HttpClient();
        }


#if GRPC

        #region GrpcHost
        private string _grpcHost;
        private string _tempGrpcHost;
        private DateTime _maxTemp_Grpc;
        public string GrpcHost {
            set {
                _grpcHost = value;
            }
            get {
#if Consul
                if (_grpcHost == null) {

                    if (string.IsNullOrEmpty(_tempGrpcHost) || DateTime.Now > _maxTemp_Grpc) {
                        var task = ConsulServiceDiscovery.DiscoveryOne(ConsulAddress, HttpServiceName);
                        task.Wait();
                        _tempGrpcHost = task.Result;
                        _maxTemp_Grpc = DateTime.Now.AddSeconds(OvertimeSeconds);
                    }
                    return _tempGrpcHost;
                }
#endif
                return _grpcHost;
            }
        }
        #endregion

        private bool isCreate;
        public TextFilterGrpcProvider CreateTextFilterGrpcClient()
        {
            if (isCreate == false) {
                AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.Http2UnencryptedSupport", true);
                isCreate = true;
            }
            return new TextFilterGrpcProvider(GrpcHost);
        }
        public TextFilterGrpcAsyncProvider CreateTextFilterGrpcAnsycClient()
        {
            if (isCreate == false) {
                AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.Http2UnencryptedSupport", true);
                isCreate = true;
            }
            return new TextFilterGrpcAsyncProvider(GrpcHost);
        }
 
#endif


        #region CreateKeywordProvider  CreateKeywordTypeProvider CreateSysProvider
        public IKeywordProvider CreateKeywordProvider()
        {
            return new KeywordProvider(this);
        }

        public IKeywordTypeProvider CreateKeywordTypeProvider()
        {
            return new KeywordTypeProvider(this);
        }

        public ISysProvider CreateSysProvider()
        {
            return new SysProvider(this);
        }
        #endregion

        #region TextFilterProvider

        public ITextFilterProvider CreateTextFilterProvider()
        {
            return new TextFilterProvider(this);
        }

        #endregion

#if Async
        public ITextFilterAsyncProvider CreateTextFilterAsyncProvider()
        {
            return new TextFilterAsyncProvider(this);
        }
#endif


#if Consul

        public string ConsulAddress { private get; set; } = "http://localhost:8500";


        public Task<List<string>> GetServiceUrls(ServiceUrlType urlType)
        {
            if (urlType == ServiceUrlType.Grpc) {
                return ConsulServiceDiscovery.Discovery(ConsulAddress, GrpcServiceName);
            }
            return ConsulServiceDiscovery.Discovery(ConsulAddress, HttpServiceName);
        }

#endif
    }

}
